「Railsのi18nを理解する」課題の提出

wata00913 opened the pull request:
確認をよろしくお願いします！🙇‍♂️

## 補足説明
- ロケールの切り替え
  - URLにロケールを含めると切り替えできるようにしました。
  - URLにロケールを含めない場合、デフォルトロケール（ja）が反映されます。
- 「[初心者がRailsプロジェクトへの初PRする前に見るチェックリスト](https://docs.komagata.org/5676)」に従い、`json`でアクセスするメソッド（`format）を削除しました。

### 気になること
 `scaffold`で生成された不要なコード（`json`でアクセスするメソッドなど）は、どのブランチで対応すべきでしょうか？

今回はi18nのブランチで対応しましたが、i18nには関連しない機能のため別のブランチで対応すべきとも考えました。ただ、別ブランチとして切るほどの量でもないので悩ましいです。
実務ではこういった場合どのような対応をするのか気になりました🤔

wata00913 committed:
ロケール辞書ファイルとヘッダー用の訳文を作成

wata00913 committed:
リクエスト間でロケールの管理をする

wata00913 committed:
ヘッダー用のロケール変数を作成

wata00913 committed:
ヘッダーの文字を国際化

wata00913 committed:
デフォルトロケールを日本語に設定

wata00913 committed:
各メモ項目のラベルを国際化

wata00913 committed:
リンクを国際化

wata00913 committed:
国際化処理を共通化。

before_actionの呼び出しメソッドが複数であるが、コードが重複しているためまとめた。

wata00913 committed:
フォームボタンを国際化

`ファイルを選択`は、ブラウザ側の文言のため未対応

wata00913 committed:
アクション成功のメッセージを国際化

wata00913 committed:
before_actionの国際化関連のメソッド名変更と辞書の階層構造を変更

辞書のmessageは、confirmやalertなどの場合も含まれるので成功用の階層(success)を作った。
また、メソッド名が不適切だったのでリネームをした。

wata00913 committed:
enの辞書に漏れがあったので追加

wata00913 committed:
削除モーダルダイアログの国際化

wata00913 committed:
リクエストでのURL管理方法にURL paramsを元にした設定を追加

wata00913 committed:
不要な`respond_to`を削除。

現状は`json`  でのアクセスが不要のため削除した。

wata00913 committed:
不要なコメントを削除

デフォルトで記載されたコメントなので削除。

wata00913 committed:
翻訳に誤りがあったため修正

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

wata00913 committed:
レビュー指摘箇所の修正：ヘッダーのロケール化処理をViewに移行

ロケール化対応はコントローラー側ではなくView側で行うのが一般的。

wata00913 committed:
レビュー指摘箇所の修正：属性のロケール化処理をViewに移行

wata00913 committed:
レビュー指摘箇所の修正：リンクのロケール化処理をViewに移行

wata00913 committed:
レビュー指摘箇所の修正：確認ダイアログのロケール化処理をViewに移行

wata00913 committed:
不要なメソッド、コールバックの削除

wata00913 committed:
レビュー指摘箇所の修正：フォームボタンのロケール化処理をViewに移行

 フォームボタンは共通のテンプレートで定義されているため、controller.action_name
 メソッドを使ってアクションごとにキーを変更させた。

wata00913 committed:
レビュー指摘箇所の修正：作成、更新、削除処理の成功メッセージのロケールを修正

DRYをやりすぎたのでシンプルに書き直しをした。

wata00913 committed:
不要なメソッド、コールバックの削除

wata00913 committed:
レビュー箇所の修正：Bookモデルの属性のロケールをhumna_attribute_nameメソッドで置き換えと
それに伴う辞書の修正

モデルの属性の翻訳は、規約に従うとメソッドで取得できてシンプルに書けるので置き換えた。

wata00913 committed:
日本語翻訳の修正

英語と統一感がなかったので合わせた

wata00913 committed:
不要な辞書のキーを削除

コミットID`9ac75c39737d0304e11af47930bf414f265fe077`での修正によって不要になった。

JunichiIto commented on app/controllers/books_controller.rb:
```diff
@@ -28,11 +28,9 @@ def create
     respond_to do |format|
       if @book.save
-        format.html { redirect_to @book, notice: 'Book was successfully created.' }
-        format.json { render :show, status: :created, location: @book }
```
jsonをなくしてhtmlだけにするなら、

- respond_to も消す
- jbuilder
のファイルも消す

という対応が必要になりますね。

ただ、プラクティスとしては「プルリクエストには課題の要件満たすために必要なdiffだけ発生している状態がベストです。」というスタンスなので、jsonは残したままの方がいいかもしれません。

wata00913 replied:
伊藤さんの仰る通りプラクティスのスタンスに従う方が良さそうなので、jsonを残すように修正しました！

JunichiIto commented on app/views/books/_form.html.erb:
```diff
@@ -12,26 +12,26 @@
 
   <div class="actions">
-    <%= form.submit %>
+    <%= form.submit t("book.submit.#{controller.action_name}") %>
```
Railsがデフォルトで見に行くキーがあるのでそれを使うと良さそうです。
「書籍を〜」の文言は入らなくなりますが、ここでは無理に入れなくても良いかも。

https://karlley.hatenablog.jp/entry/2022/07/21/082900

wata00913 replied:
カリーさんのブログが大変参考になりました、共有ありがとうございます。

- ご指摘があったように、`submit`の引数を除去してシンプルに書けました。
- また、FormBuilderの`label`メソッドもRailsがデフォルトで見に行くキーがあるので、
同様の対応をしました。

※ gem内でデフォルト定義された辞書は、モデル名をパラメータ化してたので「書籍を」の文言はそのまま残せました！

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,28 @@
+  book:
+    heading:
+      index: "書籍一覧"
+      new: "新刊"
```
新刊、というのは言葉のチョイスとして誤解を招きそうです。

https://dictionary.goo.ne.jp/word/%E6%96%B0%E5%88%8A/

wata00913 replied:
新刊だと「出版する」という意味が含まれて不適切なので、「書籍の新規作成」としました。
「書籍の」と「作成」を追加した理由は下記になります。
- 「新規」は「新規事業」のように後ろの名詞にかかることが多いので、「新規作成」としました。
-  index:の訳は「書籍」があるので、「書籍の」を追加しました。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,28 @@
+    heading:
+      index: "書籍一覧"
+      new: "新刊"
+      edit: "編集書籍"
```
編集書籍というのも日本語としてちょっと違和感がありますね。

wata00913 replied:
[「新刊」の修正](https://github.com/wata00913/fjord-books_app/pull/1#discussion_r947553064)と同様に「書籍の編集」と修正しました。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,28 @@
+      new: "書籍を新規作成する"
+      edit: "書籍を更新する"
+    message:
+      confirm: "本当によろしいですか？"
```
こういうメッセージは逆に本以外の場所でも再利用できるので、`book:`以下に置かない方がいいと思います。

wata00913 replied:
`book:`から`common`の方に移すのと、「書籍」の部分をパラメータ化することで、再利用できる形に修正ました。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,28 @@
+    message:
+      confirm: "本当によろしいですか？"
+      success:
+        create: "書籍の作成に成功しました。"
```
```suggestion
        create: "%{model}の作成に成功しました。"
```

のようにパラメータを使うことで、再利用可能な翻訳に昇格させることができます。

wata00913 replied:
ご指摘があったように、パラメータ"%{model}"を使って再利用できる形に修正しました。

JunichiIto commented on app/views/books/_form.html.erb:
```diff
@@ -12,26 +12,26 @@
 
   <div class="field">
-    <%= form.label :title %>
+    <%= form.label :title, Book.human_attribute_name(:title) %>
```
```suggestion
    <%= form.label :title %>
```

とすれば、Railsが自動的に翻訳を取りに行ってくれると思うのですがどうでしょう？

wata00913 replied:
[submitの修正](https://github.com/wata00913/fjord-books_app/pull/1#discussion_r947537958)で先にコメント書いてしまいましたが、自動的に翻訳を取りに行ってくれたので、引数を削除しました。

JunichiIto reviewed (commented):
コメントしました〜。

wata00913 committed:
bb33805で削除したformat.jsonを元に戻す

プラクティスの要件では「課題の要件を満たすために必要なdiffだけを残す状態
が望ましい」ので、そのまま残すようにする。

wata00913 committed:
特定のモデルに対応した翻訳を再利用可能な翻訳に昇格

モメッセージは共通で利用できるため、デル固有の翻訳箇所をパラメータ化して再利用可能にした。

wata00913 committed:
フォームのlabelとsubmitメソッドのテキスト指定をする引数を除去

テキストを明示的に指定しなくてもRailsが自動的に翻訳を取りに行くため、引数を除去した。

wata00913 committed:
日本語訳を修正

以下の理由から修正
・「新刊」の意味が不適切。書籍を出版するための画面ではない。
・「編集書籍」の日本語が違和感。連体助詞「の」を追加し、後ろの名詞を「編集」にした

wata00913 committed:
モデルの属性の英語翻訳を削除

gem内のデフォルト翻訳と英訳と差分がなく不要なので削除した。

JunichiIto commented on app/controllers/books_controller.rb:
```diff
@@ -28,9 +28,12 @@ def create
     respond_to do |format|
       if @book.save
-        format.html { redirect_to @book, notice: t('book.message.success.create') }
+        I18n.translate
```
質問）このメソッド呼び出しは必要なんですか？

wata00913 replied:
呼び出しは不要です。
恐らくコードをコピペした際に、間違って不要なコードも追加したと思います。
今後は、`git diff`のチェックで不用意なコードを残さないように注意します。🙇‍♂️

感想：こういった不注意によって、エラーが発生しないバグを潜ませるんだなと実感しました😓
伊藤さんのブログ「git commitする前に必ずdiffを自分でレビューするクセを付けよう」を読み直そうと思います。

JunichiIto commented on app/views/books/edit.html.erb:
```diff
@@ -1,6 +1,6 @@
-<h1>Editing Book</h1>
+<h1><%= t('book.heading.edit') %></h1>
```
```suggestion
<h1><%= t('.heading') %></h1>
```

とするのもありかも？

https://railsguides.jp/i18n.html#%E9%81%85%E5%BB%B6%E6%8E%A2%E7%B4%A2

wata00913 replied:
遅延探索でよりスッキリ書けるんですね〜
今回は伊藤さんの[アドバイス](https://github.com/wata00913/fjord-books_app/pull/1#discussion_r948595385)で対応することに決めて、commonに辞書を持っていったので遅延探索の対応は行いませんでした。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -0,0 +1,32 @@
+      new: "書籍の新規作成"
+      edit: "書籍の編集"
+    action:
+      new: "新規"
```
訳文に「書籍の」が入らないなら、bookではなくcommonに持っていって再利用可能な翻訳にしてもいいかもしれません。
また、「書籍の」が入る場合でもパラメータ化することで再利用できますね。

wata00913 replied:
「書籍の」を省いてcommonに持っていく方針にしました！（action配下の辞書も同様にcommonへ。）
今回のプラクティスは、書籍関連の画面だけなので、わざわざ「書籍の」を追加しなくても良いかな〜とも思えました。

JunichiIto reviewed (commented):
コメントしました！

wata00913 committed:
不要なメソッド呼び出しを削除。

間違って不要なコードを追加したものと思われる。

wata00913 committed:
モデル固有の辞書を再利用可能な翻訳に昇格

「書籍の」箇所以外の文言は再利用できるかつ、今回の画面は書籍に関する画面しかないため「書籍の」箇所は省略した。
また、英語も同様の対応。

JunichiIto commented on config/locales/ja.yml:
```diff
@@ -19,11 +19,10 @@ ja:
-      index: "書籍の一覧"
-      new: "書籍の新規作成"
-      edit: "書籍の編集"
+      index: "一覧"
```
好みの問題ではありますが、ページをパッと開いたときに「これは何の一覧？」とか「このフォームは何を新規作成？」とユーザーに思われそうなので、`%{name}の一覧`のようにしておいた方がいいんじゃないかなーと僕は思いました。

JunichiIto reviewed (approved):
1点コメントしましたが、技術的な指摘ではないのでこれでOKとします〜 🙆‍♂️
